### ISSUE TITLE
Drop support for HttpClientInterface

### ISSUE THREAD

**Issue by mekras** [mekras/psr7-client#3]
Drop HttpClientInterface support in favor of more functional [Httpug](http://php-http.readthedocs.org/en/latest/).

**Comment by shulard** [comment]
Hello @mekras!

You make an impressive job with your cURL client. Are you interested to update your lib and make it the default cURL client (like defined here: php-http/httplug/issues/32) :

* Drop `AuthHelper` and use `php-http/authentification`
* Use `MessageFactory` instead of `Connectors`

**Comment by sagikazarmark** [comment]
Hey @mekras 

Httplug founder here.

As @shulard mentioned, we would be happy if you would like to make your implementation an official php-http/curl-client implementation.

Ping us on [slack](http://slack.httplug.io/) if you are interested, contact us on twitter, or just reply to these comments.

**Comment by mekras** [comment]
@sagikazarmark, I would be happy. What should I do?

**Comment by sagikazarmark** [comment]
Well, @shulard added a few good points, but he unfortunately removed his comment from here.

A few changes which might be required:

- We have our own MessageFactory solution, would be nice if you could use that.
- We also have an integration test suite, it would be nice to use that eventually.
- We have our own boilerplate for repository format, but you don't have to strictly follow it (although it would be nice to follow the same rules)
- Please make sure that you are fully credited

A few questions that you should decide:

- Would you like to MOVE the repository with the full history or preserve it for historycal reasons?
- Would you like to preserve the current versioning or restart it from 1.0?
- Are you fine with `php-http/curl-client` name?
- Would you like to make required changes first, or should we move the repository first?

You could ask @joelwurtz about further things, as he also implemented a standalone client:

https://github.com/php-http/socket-client/

These are the changes I can think of now, the rest can be done later as well.

**Comment by shulard** [comment]
Yeah, sorry about the mistake I removed the comment and Github doesn't allow to cancel that kind of thing. Then @sagikazarmark just add his comment so I though it was ok...

I started today a project structure for the `curl-client` project here: https://github.com/shulard/http-plug-curl-adapter until we found your code :smiley: 
It follow the boilerplate code and is inspired by the php-http/socket-client repository. It also retrieve all the required stuff for testing.

I'll be glad to help, if you want to :smile:

**Comment by shulard** [comment]
I think some other stuff about the changes :

* `Connectors` must be updated to use https://github.com/php-http/message-factory and https://github.com/php-http/discovery
* `AuthHelper` must be removed and use the https://github.com/php-http/authentification instead

With that this client will be totally adaptable. Also I view that you are not using the `curl_reset` function available since PHP 5.5. It allow to reset all the options linked to a cURL resource. It can be useful to avoid opening a new resource for each call (for performance and optimisation).
Currently the class [CurlHttpClient](https://github.com/shulard/http-plug-curl-adapter/blob/master/src/CurlHttpClient.php) that I pushed contains that logic.

**Comment by sagikazarmark** [comment]
Thanks for adding the stuff back.

Honestly, I haven't used cURL in PHP for ages: how is authentication related to it? Are there any cURL special authentication? Or just plain basic auth?

Authentication should be handled by the authentication plugin in the plugins repo.

**Comment by shulard** [comment]
There is a cURL option to specify basic auth credentials. But I think that inside the extension, the authorisation header is just set based on that option...
We juste need to validate (by adding an authentication test maybe) that it's the only thing to do...

**Comment by sagikazarmark** [comment]
I think the less special logic contained in a client, the better it is. We have an Authentication plugin, which uses the linked authentication library.

**Comment by mekras** [comment]
@sagikazarmark 
> Would you like to MOVE the repository with the full history or preserve it for historycal reasons?

It should be preserved, at least for some time. We use this package in our internal projects. But I can copy base code into the new `php-http/curl-client` package and maintain both while migrating our projects.

**Comment by mekras** [comment]
>  AuthHelper must be removed and use the php-http/authentification instead

#6

**Comment by sagikazarmark** [comment]
Repo created, access granted to you:

https://github.com/php-http/curl-client

**Comment by mekras** [comment]
@sagikazarmark, OK. I'll start today.

**Comment by sagikazarmark** [comment]
Some more things:

- Redirection should also be handled by plugins (affected: follow_redirects, max_redirects)
- We have a cookie plugin as well (affected: use_cookies)

**Comment by sagikazarmark** [comment]
Thanks @mekras

**Comment by shulard** [comment]
Hello @mekras!

I can handle some issues if you want... Do you need to move your code the curl-client before starting the upgrade ? Or are you wanting to make the changes directly here ?

**Comment by sagikazarmark** [comment]
I have no problem with pushing repo content to the new repo as is now.

**Comment by mekras** [comment]
@shulard, I prefer to start with a blank project and copy just what is needed without legacy staff. Actually I'm already started locally.

**Comment by sagikazarmark** [comment]
:+1:

**Comment by shulard** [comment]
Ok @mekras, just let me know :smile:
